drivers/mspi/mspi_stm32_qspi.c

```diff
@@ -0,0 +1,1198 @@
+ * @return 0 on success
+ * @return A negative errno value upon failure.
+ */
+static int mspi_stm32_qspi_execute_transfer(const struct device *dev,
+					    const struct mspi_xfer_packet *packet,
+					    uint8_t access_mode)
+{
+	HAL_StatusTypeDef hal_ret;
+	struct mspi_stm32_data *dev_data = dev->data;
+
+	if (packet->dir == MSPI_RX) {
+		/* Receive the data */
+		switch (access_mode) {
+		case MSPI_ACCESS_SYNC:
+			hal_ret = HAL_QSPI_Receive(&dev_data->hmspi.qspi, packet->data_buf,
+						   HAL_QSPI_TIMEOUT_DEFAULT_VALUE);
+			break;
+		case MSPI_ACCESS_ASYNC:
+			hal_ret = HAL_QSPI_Receive_IT(&dev_data->hmspi.qspi, packet->data_buf);
+			break;
+		case MSPI_ACCESS_DMA:
+			/* DMA mode not yet implemented for QSPI */
+			LOG_ERR("DMA mode not supported yet for QSPI");
+			return -ENOTSUP;
+		default:
+			LOG_ERR("Invalid access mode: %d", access_mode);
+			return -EINVAL;
+		}
+	} else {
+		/* Transmit the data */
+		switch (access_mode) {
+		case MSPI_ACCESS_SYNC:
+			hal_ret = HAL_QSPI_Transmit(&dev_data->hmspi.qspi, packet->data_buf,
+						    HAL_QSPI_TIMEOUT_DEFAULT_VALUE);
+			break;
+		case MSPI_ACCESS_ASYNC:
+			hal_ret = HAL_QSPI_Transmit_IT(&dev_data->hmspi.qspi, packet->data_buf);
+			break;
+		case MSPI_ACCESS_DMA:
+			/* DMA mode not yet implemented for QSPI */
+			LOG_ERR("DMA mode not supported yet for QSPI");
+			return -ENOTSUP;
+		default:
+			LOG_ERR("Invalid access mode: %d", access_mode);
+			return -EINVAL;
+		}
+	}
+
+	if ((hal_ret != HAL_OK) || (access_mode == MSPI_ACCESS_SYNC)) {
+		pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+		(void)pm_device_runtime_put(dev);
+
+		if (hal_ret != HAL_OK) {
+			LOG_ERR("Failed to start %s transfer: %d",
+				packet->dir == MSPI_RX ? "receive" : "transmit", hal_ret);
+			return -EIO;
+		}
+
+		return 0;
+	}
+
+	/* For ASYNC mode, wait for IRQ completion (PM locks released in ISR) */
+	if (k_sem_take(&dev_data->sync, K_FOREVER) < 0) {
+		LOG_ERR("Failed to complete async transfer");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Read data in memory-mapped mode (XIP)
+ *
+ * Note: Write operations are NOT supported in memory-mapped mode for QSPI.
+ * Writes must use indirect mode .
+ *
+ * @param dev Pointer to the device structure
+ * @param packet Pointer to transfer packet (must be RX direction)
+ * @return 0 on success.
+ * @return A negative errno value upon failure.
+ */
+static int mspi_stm32_qspi_memory_mapped_read(const struct device *dev,
+					      const struct mspi_xfer_packet *packet)
+{
+	struct mspi_stm32_data *dev_data = dev->data;
+	int ret;
+
+	if (!mspi_stm32_qspi_is_memmap(dev)) {
+		ret = mspi_stm32_qspi_memmap_on(dev);
+		if (ret != 0) {
+			LOG_ERR("Failed to enable memory mapped mode");
+			return ret;
+		}
+	}
+
+	uintptr_t mmap_addr = dev_data->memmap_base_addr + packet->address;
+
+	/* Memory-mapped mode is READ-ONLY for QSPI */
+	LOG_DBG("Memory-mapped read from 0x%08lx, len %zu", mmap_addr, packet->num_bytes);
```

Comment:
`num_byte` is of type `uint32_t`, not `size_t`.
```suggestion
	LOG_DBG("Memory-mapped read from 0x%08lx, len %u", mmap_addr, packet->num_bytes);